server/src/main/java/org/elasticsearch/cluster/metadata/FieldInferenceMetadata.java

```diff
@@ -0,0 +1,235 @@
+    }
+
+    public FieldInferenceMetadata(MappingLookup mappingLookup) {
+        this.inferenceIdForField = ImmutableOpenMap.builder(mappingLookup.getInferenceIdsForFields()).build();
+        ImmutableOpenMap.Builder<String, Set<String>> sourcePathsBuilder = ImmutableOpenMap.builder(inferenceIdForField.size());
+        inferenceIdForField.keySet().forEach(fieldName -> sourcePathsBuilder.put(fieldName, mappingLookup.sourcePaths(fieldName)));
+        this.sourceFields = sourcePathsBuilder.build();
+    }
+
+    public FieldInferenceMetadata(StreamInput in) throws IOException {
+        inferenceIdForField = in.readImmutableOpenMap(StreamInput::readString, StreamInput::readString);
+        sourceFields = in.readImmutableOpenMap(StreamInput::readString, i -> i.readCollectionAsImmutableSet(StreamInput::readString));
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeMap(inferenceIdForField, StreamOutput::writeString);
+        out.writeMap(sourceFields, StreamOutput::writeStringCollection);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.field(INFERENCE_FOR_FIELDS_FIELD.getPreferredName(), inferenceIdForField);
+        builder.field(SOURCE_FIELDS_FIELD.getPreferredName(), sourceFields);
+
+        return builder;
+    }
+
+    @SuppressWarnings("unchecked")
+    private static final ConstructingObjectParser<FieldInferenceMetadata, Void> PARSER = new ConstructingObjectParser<>(
+        "field_inference_metadata_parser",
+        false,
+        (args, unused) -> new FieldInferenceMetadata((Map<String, String>) args[0], (Map<String, Set<String>>) args[1])
+    );
+
+    static {
+        PARSER.declareObject(ConstructingObjectParser.constructorArg(), (p, c) -> p.mapStrings(), INFERENCE_FOR_FIELDS_FIELD);
+        PARSER.declareObject(
+            ConstructingObjectParser.constructorArg(),
+            (p, c) -> p.map(
+                HashMap::new,
+                v -> v.list().stream().map(Object::toString).collect(Collectors.toSet())
+            ),
+            SOURCE_FIELDS_FIELD
+        );
+    }
+
+    public static FieldInferenceMetadata fromXContent(XContentParser parser) throws IOException {
+        return PARSER.parse(parser, null);
+    }
+
+    @Override
+    public Diff<FieldInferenceMetadata> diff(FieldInferenceMetadata previousState) {
+        return new FieldInferenceMetadataDiff(previousState, this);
+    }
+
+    public String getInferenceIdForField(String field) {
+        return inferenceIdForField.get(field);
+    }
+
+    public Map<String, String> getInferenceIdForFields() {
+        return inferenceIdForField;
+    }
+
+    public Set<String> getSourceFields(String field) {
+        return sourceFields.get(field);
+    }
+
+    public Map<String, Set<String>> getFieldsForInferenceIds() {
```

Comment:
Could this method be accessed by multiple threads? Should we mark `fieldsForInferenceIds` as `volatile` or maybe even wrap it in `AtomicReference`?

Reply:
Yes, it can be accessed by multiple threads. 

I didn't want to overdo synchronization / early optimization, and have kept it light on purpose. 

I think it's preferrable to recalculate vs having the syncing overhead - but happy to hear your thoughts.

Reply:
Makes sense, adding that

Reply:
We need to have a single version of this map that is immutable and final. It's a small map so there's no need to compute all the flavours preemptively or even to cache them. The consumer can rebuild data structures on top if their access pattern is different. 
Can we just keep a `Map<String, FieldInference>` where the keys are the semantic text fields and `FieldInference` contains the inference id and the list of source fields? 
This way it's easy to extract the FieldInference for a field query.
For ingest we need to visit the entire map anyway so any shape is ok.

Reply:
I see that this is a similar pattern to other Index Metadata related structures. Changing it